### PR TITLE
CNI plugin changes: Windows Task Networking

### PR DESCRIPTION
**Pull request by rawahars**
***Issue #, if available:***
Issue 1: Support for Windows task networking

Issue 2:
On Windows, whenever an ENI is attached to a Windows instance, a stale route entry is added in the default network namespace.
Due to this entry, we cannot access the task using the task ENI’s IP from the same instance. This also means that the task cannot be accessed from any non-awsvpc mode tasks present on the same instance.
This entry needs to be deleted from the default network namespace.
 
**Steps to replicate the issue:**
1. Launch an EC2 Windows instance
2. Create a new network interface from EC2 console
3. Connect this network interface to the Windows instance

We will find a stale route entry as below:
| ifIndex  | DestinationPrefix | NextHop |
| ------------- | ------------- | ------------- |
| ifIndex of the ENI   | ENI IP Address/32  | 0.0.0.0 |            

***Description of changes:***
1. Addition of AWSVPC Capability for vpc-shared-eni plugin
2. When vpc-shared-eni CNI plugin is invoked for task networking then on invocation of DEL, we will delete l2bridge network created for Task ENI
3. The changes are introduced to delete the stale route entry of the task ENI from the host compartment's routing table.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: network/netutils/netutils.go [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package netutils
+
+import "net"
+
+// NetUtils is a wrapper for executing Windows networking commands
+type NetUtils interface {
```
Comment (vsiddharth) [inline network/netutils/netutils.go:19]: Should this be renamed to RouteUtils

File: network/netutils/netutils_route_windows.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	return &netUtils{}
+}
+
+// AddRoute adds a route entry in the routing table
+func (netsh *netUtils) AddRoute(destination *net.IP, subnetMask *net.IP, gateway *net.IP, interfaceNum *int) error {
+	if destination == nil || subnetMask == nil || gateway == nil {
+		return errors.New("unable to delete route due to invalid arguments")
+	}
+
+	commandArgs := netsh.generateCommandArgs(destination, subnetMask, gateway, interfaceNum)
```
Comment (vsiddharth) [inline network/netutils/netutils_route_windows.go:43]: Can you log the command args?

File: network/netutils/netutils_route_windows.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+	return exec.Command(routeAddCommand, commandArgs...).Run()
+}
+
+// DeleteRoute deletes a route entry from the routing table
+func (netsh *netUtils) DeleteRoute(destination *net.IP, subnetMask *net.IP, gateway *net.IP) error {
+	if destination == nil || subnetMask == nil {
+		return errors.New("unable to delete route due to invalid arguments")
+	}
+
+	commandArgs := netsh.generateCommandArgs(destination, subnetMask, gateway, nil)
```
Comment (vsiddharth) [inline network/netutils/netutils_route_windows.go:53]: Same comment as above

File: plugins/vpc-shared-eni/config/netconfig.go [hunk 4]
```diff
@@ -40,23 +40,30 @@ type NetConfig struct {
...
+	ENIIPAddress     string        `json:"eniIPAddress"`
+	VPCCIDRs         []string      `json:"vpcCIDRs"`
+	BridgeType       string        `json:"bridgeType"`
+	BridgeNetNSPath  string        `json:"bridgeNetNSPath"`
+	IPAddress        string        `json:"ipAddress"`
+	GatewayIPAddress string        `json:"gatewayIPAddress"`
+	InterfaceType    string        `json:"interfaceType"`
+	TapUserID        string        `json:"tapUserID"`
+	ServiceCIDR      string        `json:"serviceCIDR"`
+	TaskENI          TaskENIConfig `json:"taskENI"`
```
Comment (vsiddharth) [inline plugins/vpc-shared-eni/config/netconfig.go:61]: nit: rename `taskENI` to `taskENIConfig`

File: plugins/vpc-shared-eni/config/netconfig.go [hunk 5]
```diff
@@ -40,23 +40,30 @@ type NetConfig struct {
...
+}
+
+// TaskENI defines the Task Networking specific data required by the plugin
```
Comment (vsiddharth) [inline plugins/vpc-shared-eni/config/netconfig.go:64]: Fix comment

File: plugins/vpc-shared-eni/config/netconfig.go [hunk 6]
```diff
@@ -40,23 +40,30 @@ type NetConfig struct {
...
+type TaskENIConfig struct {
+	Enable bool `json:"enableTaskENI"`
```
Comment (vsiddharth) [inline plugins/vpc-shared-eni/config/netconfig.go:66]: `Enable bool` is not conveying intent clearly. 
Can this be renamed to `PauseContainer` to indicate the infra container in use?

File: plugins/vpc-shared-eni/network/bridge_windows.go [hunk 7]
```diff
@@ -34,6 +36,9 @@ const (
 
 	// hnsEndpointNameFormat is the format of the names generated for HNS endpoints.
 	hnsEndpointNameFormat = "cid-%s"
+
+	// loopbackRouteEntryMask is the mask of loopback entry in routing table
+	loopbackRouteEntryMask = "255.255.255.255"
```
Comment (vsiddharth) [inline plugins/vpc-shared-eni/network/bridge_windows.go:41]: This is not related to loopback. Please rectify. 
A subnet mask of `255.255.255.255` puts each device inside its own subnet and generally indicates a point to point link.

File: plugins/vpc-shared-eni/network/bridge_windows.go [hunk 8]
```diff
@@ -263,6 +268,17 @@ func (nb *BridgeBuilder) FindOrCreateEndpoint(nw *Network, ep *Endpoint) error {
 	// Return network interface MAC address.
 	ep.MACAddress, _ = net.ParseMAC(hnsResponse.MacAddress)
 
+	// For task-networking, we need to remove the stale route entries from the routing table
```
Comment (vsiddharth) [inline plugins/vpc-shared-eni/network/bridge_windows.go:271]: Add `default host compartment`.

File: plugins/vpc-shared-eni/network/bridge_windows.go [hunk 9]
```diff
@@ -263,6 +268,17 @@ func (nb *BridgeBuilder) FindOrCreateEndpoint(nw *Network, ep *Endpoint) error {
...
+	// Windows creates a loopback entry for each ENI by default which needs to be removed
```
Comment (vsiddharth) [inline plugins/vpc-shared-eni/network/bridge_windows.go:272]: This should not be called as loopback entry.

File: plugins/vpc-shared-eni/network/bridge_windows.go [hunk 10]
```diff
@@ -263,6 +267,16 @@ func (nb *BridgeBuilder) FindOrCreateEndpoint(nw *Network, ep *Endpoint) error {
 	// Return network interface MAC address.
 	ep.MACAddress, _ = net.ParseMAC(hnsResponse.MacAddress)
 
+	// For task-networking, we need to remove the stale route entries from the routing table of default host compartment
+	if nw.TaskENIConfig.PauseContainer {
+		mask := net.ParseIP(staticRouteEntryMask)
+		err := routeutils.New().DeleteRoute(&ep.IPAddress.IP, &mask, nil)
```
Comment (vsiddharth) [inline plugins/vpc-shared-eni/network/bridge_windows.go:273]: What happens if multiple attempts are made to delete the same route?
Would it be better to delete only if the required route is present?

Comment (harsh-rawat-a2z) [reply plugins/vpc-shared-eni/network/bridge_windows.go:275]: The route entry will be deleted only if it is present in the routing table of the host compartment. Otherwise nothing happens.
Error is returned only when the route command itself is not able to execute.

File: network/routeutils/routeutils_windows.go [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+// AddRoute adds a route entry in the routing table
+func (utils *routeUtils) AddRoute(destination *net.IP, subnetMask *net.IP, gateway *net.IP, interfaceNum *int) error {
+	if destination == nil || subnetMask == nil || gateway == nil {
+		return errors.New("unable to delete route due to invalid arguments")
+	}
+
+	commandArgs := utils.generateCommandArgs(destination, subnetMask, gateway, interfaceNum)
+	log.Infof("Executing command: %v", commandArgs)
+
+	return exec.Command(routeAddCommand, commandArgs...).Run()
```
Comment (mogren) [inline network/routeutils/routeutils_windows.go:48]: What is the exact command here. Will it fail if the route already exists? There was some issues before in the Linux CNI where `netlink.RouteAdd` failed since the route was already existing, so we switched all calls to be `netlinkRouteReplace`.

**Comment by ofiliz** [comment]
Closing this PR because we agreed that it is no longer needed in our offline discussion.
